mesonbuild/interpreter/interpreter.py

```diff
@@ -2076,6 +2078,29 @@ def func_install_symlink(self, node: mparser.BaseNode,
         self.build.symlinks.append(l)
         return l
 
+    @FeatureNew('structured_sources', '0.62.0')
+    @typed_pos_args('structured_sources', object, optargs=[dict])
+    @noKwargs
+    @noArgsFlattening
+    def func_structured_sources(
+            self, node: mparser.BaseNode,
+            args: T.Tuple[object, T.Optional[T.Dict[str, object]]],
+            kwargs: 'TYPE_kwargs') -> build.StructuredSources:
+        valid_types = (str, mesonlib.File, build.GeneratedList, build.CustomTarget, build.CustomTargetIndex, build.GeneratedList)
+        sources: T.Dict[str, T.List[T.Union['mesonlib.FileOrString', 'build.GeneratedTypes']]] = collections.defaultdict(list)
+
+        for arg in mesonlib.listify(args[0]):
+            if not isinstance(arg, valid_types):
+                raise InvalidArguments(f'structured_sources: type "{type(arg)}" is not valid')
+            sources[''].append(arg)
```

Comment:
Good place for the constant suggested above.

Reply:
I did away with one magical use with a method, so this is the only use left. I'm disinclined to add a constant for a single use, what do you think?

Reply:
I'll leave it up to others :).

I was envisioning something like

```python
class StructuredSource:
    ROOT_MODULE = ''

    def first_file(self) -> T.Union[str, File, CustomTarget, CustomTargetIndex, GeneratedList]:
        """Get the first source in the root
        :return: The first source in the root
        """
        return self.sources[ROOT_MODULE][0]
```